File: C/Cap08/c08ex02.c
// c08ex02.c

#include <stdio.h>
#include <stdbool.h>
#include <ctype.h>
#include <string.h>

bool validaNumero(const char* Numero) {
  int i;
  int tamanho = strlen(Numero);
  bool pontoEncontrado = false;

  for (i = 0; i <= tamanho - 1; ++i) {
    if (Numero[i] == '.') {
      if (pontoEncontrado) {
        return false;
      }
      pontoEncontrado = true;
    } else if (!(isdigit(Numero[i]))) {
      return false;
    }
  }

  return true;
}


int main(void) {
  char Numero[100];

  do {
    printf("Entre um valor numerico: ");
    fgets(Numero, sizeof(Numero), stdin);
    Numero[strcspn(Numero, "\n")] = '\0';
    printf("\n");


    if (Numero[0] == '\0') {
      printf("Erro: Tecla <Enter> acionada acidentalmente.\n");
      printf("\n");
      continue;
    }

    if (!validaNumero(Numero)) {
      printf("Erro: Entre um numero	valido.\n");
      printf("\n");
      continue;
    }

    break;
  } while (true);

  printf("Entrada valida fornecida: %s\n", Numero);

  printf("\n");
  printf("Tecle <Enter> para encerrar... ");
  while (getchar() != '\n');

  return 0;
}

File: DART/Cap08/c08ex02.dart
// c08ex02.dart

import 'dart:io';

bool validaNumero(String entrada) {
  final RegExp padrao = RegExp(r'^[+-]?(\d+|\d+\.\d+)$');
  return padrao.hasMatch(entrada);
}

void main() {
  String numero;

  while (true) {
    stdout.write("Entre um valor numerico inteiro: ");
    numero = stdin.readLineSync() ?? "";
    print("");

    if (numero.isEmpty) {
      print("Erro: Tecla <Enter> acionada acidentalmente.\n");
      continue;
    }

    if (!validaNumero(numero)) {
      print("Erro: Entre um valor numerico.\n");
      continue;
    }

    print("Entrada valida fornecida: $numero");
    break;
  }

  stdout.write("\nTecle <Enter> para encerrar... ");
  stdin.readLineSync();
}
